server/core/src/logic/transaction_operations.rs

```diff
@@ -0,0 +1,391 @@
+            currency_type_id: TEST_ID,
+            recurrence_type_id: TEST_ID,
+            predefined_expense_id: None,
+        };
+        let mock_currency_type = currency_type::Model {
+            id: TEST_ID,
+            abbreviation: TEST_STR.to_string(),
+            name: TEST_STR.to_string(),
+        };
+        let mock_transaction = transaction::Model {
+            id: TEST_ID,
+            donor_name: TEST_STR.to_string(),
+            value: test_decimal(),
+            date: NaiveDate::MIN,
+            currency_type_id: TEST_ID,
+            expense_id: TEST_ID,
+        };
+        let conn = MockDatabase::new(DatabaseBackend::MySql)
+            .append_query_results(vec![vec![mock_expense.clone()]])
+            .append_query_results(vec![vec![mock_currency_type.clone()]])
+            .append_exec_results(vec![MockExecResult {
+                last_insert_id: TEST_ID,
+                rows_affected: 1,
+            }])
+            .append_query_results(vec![vec![mock_transaction.clone()]])
+            .into_connection();
+
+        let saved_transaction_id = create_transaction(
+            &conn,
+            TEST_ID,
+            NewTransactionRequest {
+                donor_name: TEST_STR.to_string(),
+                currency_type_id: TEST_ID,
+                value: test_decimal(),
+                date: TEST_DATE.to_string(),
+                expense_id: TEST_ID,
+            },
+        )
+        .await;
+
+        check!(saved_transaction_id == Ok(TEST_ID));
+    }
+
+    #[tokio::test]
+    async fn create_transaction_error_cases() {
+        let mock_expense = expense::Model {
+            id: TEST_ID,
+            name: TEST_STR.to_string(),
+            description: TEST_STR.to_string(),
+            value: test_decimal(),
+            start_date: NaiveDate::MIN,
+            user_id: TEST_ID,
+            currency_type_id: TEST_ID,
+            recurrence_type_id: TEST_ID,
+            predefined_expense_id: None,
+        };
+        let mock_currency_type = currency_type::Model {
+            id: TEST_ID,
+            abbreviation: TEST_STR.to_string(),
+            name: TEST_STR.to_string(),
+        };
+        let conn = MockDatabase::new(DatabaseBackend::MySql)
+            // invalid expense id
+            .append_query_results(vec![Vec::<expense::Model>::new()])
+            .append_query_results(vec![vec![mock_currency_type.clone()]])
+            // invalid currency type id
+            .append_query_results(vec![vec![mock_expense.clone()]])
+            .append_query_results(vec![Vec::<currency_type::Model>::new()])
+            // unauthorized
+            .append_query_results(vec![vec![mock_expense.clone()]])
+            .append_query_results(vec![vec![mock_currency_type.clone()]])
+            // date cannot be parsed
+            .append_query_results(vec![vec![mock_expense.clone()]])
+            .append_query_results(vec![vec![mock_currency_type.clone()]])
```

Comment:
In the long run, its probably better to extract all _append_query_results_ into a chainable function for each operation method. (Of course it is not simple when the number of calls vary).

It would hide how many DB transaction a given operation needs, and the tests could remain more concise. This _append_query_results_ is on a different abstraction level than our tests.

Reply:
Oooooh very good idea if I understand it correctly, let's talk about this before the next refactor round. The abstraction level part feels beyond me I would love to hear what that means.

Reply:
We will get there, glad you like it.